lib/binance/futures/rest/http_client.ex

```diff
@@ -0,0 +1,173 @@
+    HTTPoison.get("#{@endpoint}#{url}", headers)
+    |> parse_response
+  end
+
+  def delete_binance(url, headers \\ []) when is_list(headers) do
+    HTTPoison.delete("#{@endpoint}#{url}", headers)
+    |> parse_response
+  end
+
+  def get_binance(url, params, config) do
+    case prepare_request(:get, url, params, config, true) do
+      {:error, _} = error ->
+        error
+
+      {:ok, url, headers} ->
+        get_binance(url, headers)
+    end
+  end
+
+  def delete_binance(url, params, config) do
+    case prepare_request(:delete, url, params, config, true) do
+      {:error, _} = error ->
+        error
+
+      {:ok, url, headers} ->
+        delete_binance(url, headers)
+    end
+  end
+
+  def post_binance(url, params, config, signed? \\ true) do
+    case prepare_request(:post, url, params, config, signed?) do
+      {:error, _} = error ->
+        error
+
+      {:ok, url, headers, body} ->
+        case HTTPoison.post("#{@endpoint}#{url}", body, headers) do
+          {:error, err} ->
+            {:error, {:http_error, err}}
+
+          {:ok, %{status_code: status_code} = response} when status_code not in 200..299 ->
+            case Poison.decode(response.body) do
+              {:ok, %{"code" => code, "msg" => msg}} ->
+                {:error, {:binance_error, %{code: code, msg: msg}}}
+
+              {:error, err} ->
+                {:error, {:poison_decode_error, err}}
+            end
+
+          {:ok, response} ->
+            case Poison.decode(response.body) do
+              {:ok, data} -> {:ok, data}
+              {:error, err} -> {:error, {:poison_decode_error, err}}
+            end
+        end
+    end
+  end
+
+  def put_binance(url, params, config, signed? \\ true) do
+    case prepare_request(:put, url, params, config, signed?) do
+      {:error, _} = error ->
+        error
+
+      {:ok, url, headers, body} ->
+        case HTTPoison.put("#{@endpoint}#{url}", body, headers) do
+          {:error, err} ->
+            {:error, {:http_error, err}}
+
+          {:ok, %{status_code: status_code} = response} when status_code not in 200..299 ->
+            case Poison.decode(response.body) do
+              {:ok, %{"code" => code, "msg" => msg}} ->
+                {:error, {:binance_error, %{code: code, msg: msg}}}
+
+              {:error, err} ->
+                {:error, {:poison_decode_error, err}}
+            end
+
+          {:ok, %{body: ""}} ->
+            {:ok, ""}
+
+          {:ok, response} ->
+            case Poison.decode(response.body) do
+              {:ok, data} -> {:ok, data}
+              {:error, err} -> {:error, {:poison_decode_error, err}}
+            end
+        end
+    end
+  end
+
+  defp prepare_request(method, url, params, config, signed?) do
+    case validate_credentials(config) do
+      {:error, _} = error ->
+        error
+
+      {:ok, %Config{api_key: api_key, api_secret: api_secret}} ->
+        cond do
```

Comment:
This function has 44 lines, which makes it too complex and hard to read. Consider using functions with pattern matching instead of that `cond`